Add section in event filter reference for JS functions and add new SQE to SQE reference (#2640)
sensu/sensu-docs: content/sensu-go/6.0/reference/filters.md

`event.entity.system.platform_version` | string  | The entity’s operating system version
`event.entity.user`                    | string  | Sensu [RBAC][25] username used by the agent entity

## Build event filter expressions with JavaScript execution functions

**COMMERCIAL FEATURE**: Access built-in JavaScript event filter execution functions in the packaged Sensu Go distribution.
For more information, see [Get started with commercial features][44].

In addition to [Sensu query expressions][27], Sensu includes several built-in JavaScript functions for event filter execution:

- `sensu.FetchEvent`
- `sensu.CheckStatus`
- `sensu.ListEvents`

Use these functions to query your event stores for other events in the same namespace.

For example, to handle only events for the `server01` entity *and* the `disk` check, use the `sensu.FetchEvent` function in your event filter expressions:

{{< code javascript >}}
"expressions": [

>   '(function () { var diskEvent = sensu.FetchEvent("router01", "disk"); if (diskEvent == nil) { return false; } return diskEvent.check.status == 0; })()'

Reviewer: You used `server01` as the entity name in the example above, might be better if we stay consistent.

]
{{< /code >}}

### `sensu.EventStatus`

The `sensu.EventStatus` function takes zero or more checks as arguments.
It returns an array of status codes for the events associated with the specified checks.

If you do not specify any checks, the function always returns an empty array.

You can refer to the checks as strings:

{{< code javascript >}}
sensu.EventStatus("database", "disk")
{{< /code >}}

If you pass the check names as strings, Sensu assumes that the entities are the same as those in the events being filtered.

You can also refer to the checks in objects that include both the entity and check name.
For example:

{{< code javascript >}}
sensu.EventStatus({entity: "server01", check: "disk"}, {entity: "server01", check: "database"})
{{< /code >}}

In both cases, if no event matches the specified entities and checks, Sensu will raise an error.

### `sensu.FetchEvent`

> The `sensu.FetchEvent` function loads the Sensu events that correspond to the specified entity and check names.

Reviewer 2: `FetchEvent` loads only a single event.

> The `sensu.FetchEvent` function loads the Sensu events that correspond to the specified entity and check names.

Author: ```suggestion
The `sensu.FetchEvent` function loads the Sensu event that corresponds to the specified entity and check names.
```

You can refer to the checks as strings:

{{< code javascript >}}
sensu.EventStatus("database", "disk")
{{< /code >}}

If you pass the check names as strings, Sensu assumes that the entites are the same as those in the events being filtered.

You can also refer to the checks in objects that include both the entity and check name.
For example:

{{< code javascript >}}
sensu.EventStatus({entity: "server01", check: "disk"}, {entity: "server01", check: "database"})
{{< /code >}}

In both cases, if no event matches the specified entities and checks, Sensu will raise an error.

### `sensu.FetchEvent`

The `sensu.FetchEvent` function loads the Sensu event that corresponds to the specified entity and check names.

The format is `sensu.FetchEvent(entity, check)`.
For example:

{{< code javascript >}}
sensu.FetchEvent("server01", "disk")
{{< /code >}}

You can only load events from the same namespace as the event being queried.
The returned object uses the same format as responses for the [events API][43].

If an event does not exist for the specified entity and check names, Sensu will raise an error.

### `sensu.ListEvents`

> The `sensu.ListEvents` function returns an array of all events in the specified namespace.

Author: What if you do not specify a namespace?

Reviewer 2: Sorry, that was poorly worded. The namespace cannot be supplied by the user. It is automatically determined by the event being filtered.

{{< /language-toggle >}}

## Use built-in JavaScript filter execution functions

Sensu includes four built-in JavaScript functions for event filter execution:

- `sensu.CheckDependencies`
- `sensu.FetchEvent`
- `sensu.CheckStatus`
- `sensu.ListEvents`

> You can use these functions to query your event stores.

Author: ```suggestion
You can use these functions to query your event stores.
For example, the following event filter allows handling only for events for the `server01` entity *and* the `disk` check:

{{< language-toggle >}}

{{< code yml >}}
type: EventFilter
api_version: core/v2
metadata:
  name: server01_disk_events
  namespace: default
spec:
  action: allow
  expressions:
  - sensu.FetchEvent("server01", "disk")
{{< /code >}}

{{< code json >}}
{
  "type": "EventFilter",
  "api_version": "core/v2",
  "metadata": {
    "name": "server01_disk_events",
    "namespace": "default"
  },
  "spec": {
    "action": "allow",
    "expressions": [
      sensu.FetchEvent("server01", "disk")
    ]
  }
}
{{< /code >}}

{{< /language-toggle >}}
```

### `sensu.CheckDependencies`

The `sensu.CheckDependencies` function takes zero or more checks as arguments.
It returns `true` if all the specified checks are passing or `false` if any of the specified checks are failing.

If you do not specify any checks, the function always returns `true`.

You can refer to the checks as strings:

{{< code javascript >}}
sensu.CheckDependencies("database", "disk")
{{< /code >}}

> 

Author: I'm thinking it would be better to provide these examples in an actual filter expression...

Reviewer 2: So, to be clear, `sensu.CheckDependencies("database", "disk")` is a full filter expression on its own.

Reviewer 2: The others are more complex. We don't intend for most people to use them, and are really only provided for tinkering.

Author: I'm sorry, I wasn't clear and also may not have been thinking about this clearly. I was wondering if we shouldn't provide a complete example filter definition that includes one of these functions as the expression.

Author: I added an example above of what I was thinking. Does this make sense at all? Or is it just more confusing? edit: also I'm not sure my example is correct in any case.

Reviewer 2: Sorry, my earlier comment was wrong.

No, this does not make sense. Sensu Query Expressions have to return a boolean. `sensu.FetchEvent` returns a complex data structure (an event).

You'd have to do something like
```
"expressions": [
  '(function () { var diskEvent = sensu.FetchEvent("router01", "disk"); if (diskEvent == nil) { return false; } return diskEvent.check.status == 0; })()'
]
```

Reviewer 2: I don't believe these functions belong inlined in the Sensu Query Expressions docs. They are building blocks for creating Sensu Query Expressions (except for sensu.CheckDependencies, which is a Sensu Query Expression, because it returns a boolean)

Author: ^ I have them in the event filter reference rather than the SQE reference. Is that what you mean?

You can also refer to the checks in objects that include both the entity and check name.
For example:

{{< code javascript >}}
sensu.CheckDependencies({entity: "server01", check: "disk"}, {entity: "server01", check: "database"})
{{< /code >}}

In both cases, if no event matches the specified entities and checks, Sensu will raise an error.

### `sensu.EventStatus`

The `sensu.EventStatus` function takes zero or more checks as arguments.
It returns an array of status codes for the events associated with the specified checks.

You can refer to the checks as strings:

{{< code javascript >}}
sensu.EventStatus("database", "disk")
{{< /code >}}

If you pass the check names as strings, Sensu assumes that the entites are the same as those in the events being filtered.

You can also refer to the checks in objects that include both the entity and check name.
For example:

{{< code javascript >}}
sensu.EventStatus({entity: "server01", check: "disk"}, {entity: "server01", check: "database"})
{{< /code >}}

In both cases, if no event matches the specified entities and checks, Sensu will raise an error.

### `sensu.FetchEvent`

> The `sensu.FetchEvent` function loads the Sensu event that corresponds to the specified entity and check names.

Author: ```suggestion
The `sensu.FetchEvent` function loads the Sensu events that correspond to the specified entity and check names.
```

The format is `sensu.FetchEvent(entity, check)`.
For example:

{{< code javascript >}}
sensu.FetchEvent("server01", "disk")
{{< /code >}}

> You can only load events from the same namespace as the event being queried.

Author: Something seems not quite right with this statement. Should it be "the entity being queried" instead? As written, it seems circular maybe (...events from...the event being queried).

Reviewer 2: That should say "the event being filtered"

Author: ohhh right I see heh. I was trying to avoid using "filtered" in a general sense but we are actually talking about the subject event for the event filter. I get it now, thank you.

> You can only load events from the same namespace as the event being queried.

Author: ```suggestion
You can only load events from the same namespace as the event being filtered.
```

The returned object uses the same format as responses for the [events API][43].

If an event does not exist for the specified entity and check names, Sensu will raise an error.

### `sensu.ListEvents`

> The `sensu.ListEvents` function returns an array of all events in the same namespace as the event being queried.

Author: ```suggestion
The `sensu.ListEvents` function returns an array of all events in the same namespace as the event being filtered.
```

{{< /language-toggle >}}

## Use built-in JavaScript filter execution functions

Sensu includes four built-in JavaScript functions for event filter execution:

- `sensu.CheckDependencies`
- `sensu.FetchEvent`
- `sensu.CheckStatus`
- `sensu.ListEvents`

You can use these functions to query your event stores.

### `sensu.CheckDependencies`

The `sensu.CheckDependencies` function takes zero or more checks as arguments.
It returns `true` if all the specified checks are passing or `false` if any of the specified checks are failing.

> If you do not specify any checks, the function returns `true`.

Author: Is there a reason to specify no checks? What would that tell you?

Reviewer 2: No, it's just the base case of the function.

You can refer to the checks as strings:

{{< code javascript >}}
sensu.CheckDependencies("database", "disk")
{{< /code >}}

If you pass the check names as strings, Sensu assumes that the entites are the same as those in the events being filtered.

You can also refer to the checks in objects that include both the entity and check name.
For example:

{{< code javascript >}}
sensu.CheckDependencies({entity: "server01", check: "disk"}, {entity: "server01", check: "database"})
{{< /code >}}

In both cases, if no event matches the specified entities and checks, Sensu will raise an error.

### `sensu.EventStatus`

> The `sensu.EventStatus` function takes zero or more checks as arguments.

Author: Is it correct that this takes checks as arguments?

Reviewer 2: See the examples for the spec of what the function takes. It can either take string check names, or objects that specify an entity name and check name. If only the check name is given, then the entity is computed from the event that is being filtered.

Author: I figured that but the gist just said it takes the same arguments as sensu.CheckDependencies so I wanted to make sure.
